src/schemas/appSettings.ts

```diff
@@ -0,0 +1,88 @@
+import { z } from 'zod'
+
+export const RegularProxySettingsSchema = z.object({
+  mode: z.literal('regular'),
+  port: z
+    .number({ message: 'Port number is required' })
+    .int()
+    .min(1)
+    .max(65535, { message: 'Port number must be between 1 and 65535' }),
+  automaticallyFindPort: z.boolean(),
+})
+
+export const UpstreamProxySettingsSchema = RegularProxySettingsSchema.extend({
+  mode: z.literal('upstream'),
+  url: z.string().url({ message: 'Invalid URL' }).or(z.literal('')),
+  requiresAuth: z.boolean(),
+  username: z.string().optional(),
+  password: z.string().optional(),
+})
```

Comment:
small nit: I would have probably created a `BaseProxySettingsSchema` to be shared between all the proxy modes and then have each of them extend that since defining the port is a common setting 🤔  (but it does not matter much here and can be a future refactor 🙇 )